### PR TITLE
Add public API for media upload with async support

### PR DESCRIPTION
**Pull request by jonmast**
Create public methods for Twitter's media upload API. These accept additional parameters, including the media_category parameter to enable asynchronous uploads. The old upload_with_media method now calls these new methods under the hood.

This is a bit more low-level than #805, but could potentially solve the same use case. Adding support for async video upload was the driving concern though.

### CODE REVIEW

**Comment by jonmast** [comment]
Just noticed #796 is also trying to solve the same problem, but has a less complete solution.

**Comment by jonmast** [comment]
I've update this to improve the documentation and satisfy yardstick. I am curious why yardstick is set to require the '@api' tag, but none of the code actually uses it.

File: lib/twitter/rest/media.rb [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+      end
+
+      # Initialize a chunked upload
+      #
+      # @param file [File] Media file being uploaded
+      # @param options [Hash] Additional parameters
+      def chunked_upload_init(file, options)
+        Twitter::REST::Request.new(self, :post, 'https://upload.twitter.com/1.1/media/upload.json',
+                                   command: 'INIT',
+                                   media_type: 'video/mp4',
```
Comment (tak1n) [inline lib/twitter/rest/media.rb]: Without having `media_category: 'tweet_video'` specified it uses the "sync" upload.

This means videos are restricted to a duration between 0:30 and 1:00.
U can read more restrictions up here: https://dev.twitter.com/rest/media/uploading-media#videorecs

Without setting this key I get following error:
```
2017-01-11T11:23:20.793Z 32352 TID-gom2y8e00 WARN: Twitter::Error::BadRequest: Duration too long, maximum:30000, actual:60095 (MediaId: snf:819142603191054338)
```

When adding the `media_category: 'tweet_video'` I was able to upload the video which is slightly longer than 1 minute (60095ms).

**Comment by tak1n** [comment]
I left a comment about `media_category` option.

Other than that I tested it to upload normal images and gifs which still works. The code seems also quite readable for me.

Anything else missing @sferik ?

**Comment by stve** [comment]
@jonmast thanks for working on this!

👍  What you've put in place looks really good, I prefer this solution to #805. 

A couple of quick thoughts/suggestions:

I think it might make sense to try to define the `media_category` as @tak1n suggested as that seems to impact success when uploading. I was able to find a discussion where the list is enumerated: https://twittercommunity.com/t/media-category-values/64781/6 It'd be nice if this were better documented by Twitter but lacking documentation, that forum discussion will have to suffice.

I'm not sure I'm keen on adding `upload_media_simple` and `upload_media_chunked` to the public API. Instead, I wonder if it would be better to change `upload` from private to public instead and hiding the details of simple/chunked behind a single public method. 

@sferik i'd be curious as to your thoughts on this as well.

**Comment by jonmast** [comment]
@tak1n I am aware of the change in behavior with the `media_category` parameter, that was the reason for implementing this. I'll see if I can add some additional documentation for this parameter to make it clearer.

@stve I agree that `upload_media_simple` and `upload_media_chunked` are a bit low-level. The current heuristic in the `upload` method feels like a bit of a hack to me though, perhaps we could use file size instead?

File: lib/twitter/rest/media.rb (hunk 1 above)
Comment (tak1n) [inline lib/twitter/rest/media.rb]: @jonmast sorry missed that options param :smile:

**Comment by stve** [comment]
> The current heuristic in the upload method feels like a bit of a hack to me though, perhaps we could use file size instead?

```ruby
if !(File.basename(media) =~ /\.mp4$/)
```

You are referring to this, yes? If so, I completely agree. 

We actually have 2 concerns: which upload method to use, and which `media_category` to specify. For upload method, I think you are correct that file size is the appropriate gauge to use. For `media_category` it seems that length and file type are the primary factors. I think we could make sensible defaults for `media_category` based on the file extension and allow it to be overridden by the user if they so choose. Alternatively, we could take this up to a higher abstraction level and create convenience methods for various media types (`upload_video`, `upload_animated_gif`, etc.)

**Comment by RyanAmos** [comment]
What's holding this PR from being merged in outside of the tests being fixed?

**Comment by jonmast** [comment]
I've been running this in production and occasionally exceeded the status checks for long videos. I've doubled the number of checks in an effort to address this.

**Comment by pas256** [comment]
It sounds like this is good to be merged now. Can we?

**Comment by stve** [comment]
@jonmast after giving this some more thought, it is probably best to leave the `media_category` as an option that the user can choose to define. 

But I would like to expose publicly a single `upload` method and make the `upload_media_simple` and `upload_media_chunked` private, using file size to determine when to use the appropriate method. 

Happy to merge once those changes are in place.

**Comment by jonmast** [comment]
@stve I had forgotten about this PR, I've updated it with the changes you suggested.

CI appears to be failing due to the rubocop issues fixed in https://github.com/sferik/twitter/pull/868

**Comment by renatolond** [comment]
Hi!

I was testing this PR because I have a gif that was converted to mp4 in another social network (Mastodon) and in my use case I'm posting it to Twitter. The thing is that according to twitter:

> Upload the media file(s) using either the simple (image only)

And since the code is checking for size to choose between chunked or simple, it was choosing only simple for my case. (Since the gif is quite small, 111169 bytes) and twitter was rejecting it (for not being an image).
I forced the chunked option and it uploaded correctly.
So I was wondering if maybe could be an option to override the size check and do chunked anyway?

Best regards

**Comment by jonmast** [comment]
@renatolond Okay, if I am understanding you correctly, your issue is that you have a very small video that is trying to upload via the "simple" endpoint due to the small filesize.

Ideally we would improve our heuristic to use the correct endpoint in all cases, the user shouldn't need to specify that. I wonder if we could perhaps use chunked endpoint for all uploads, Twitter seems to be pushing that as the preferred method going forward. Alternatively, we could check the file type and/or media_category parameter to identify the correct endpoint.

**Comment by renatolond** [comment]
That's indeed my issue, @jonmast. :)

In my case, detecting it from the file type or media_category would solve my problem.
But I guess that always doing chunked also would :)

**Comment by CedricBm** [comment]
Hey guyrs, any news on this PR? I would very much like to use this feature. :)

**Comment by kalemi19** [comment]
Any update on this?

**Comment by dekaikiwi** [comment]
@jonmast Big fan of this PR. Would be keen to see it merged! Although curious about the reasoning behind only returning the `media_id` field on a successful upload.

Other fields from the response would be quite useful to have too in the case that someone might want to cache the `media_id` for some time.

**Pull request closed without merging**